File: Data Structures & Algorithms/max-area-of-island/submission-1.cpp
class Solution {
public:
    int dfs(vector<vector<int>>& grid, int r, int c) {
        if (r < 0 || c < 0 || r >= grid.size() || c >= grid[0].size() || !grid[r][c]) {
            return 0;
        }

        grid[r][c] = 0;

        int area = 0;
        area += 1 + dfs(grid, r, c+1) + dfs(grid, r+1, c) + dfs(grid, r, c-1) + dfs(grid, r-1, c);

        return area;
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int maxArea = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == 1) {
                    int area = dfs(grid, i, j);
                    maxArea = max(maxArea, area);
                }
            }
        }

        return maxArea;
    }
};

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-1.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        rows, cols = len(matrix), len(matrix[0])
        dp = {} # (r, c) -> maxLen

        def dfs(r, c, prevVal):
            if r < 0 or r == rows or c < 0 or c == cols or matrix[r][c] <= prevVal:
                return 0
            
            if (r, c) in dp:
                return dp[(r, c)]
            
            res = 1 # min length would be one nig
            res = max(res, 1 + dfs(r + 1, c, matrix[r][c]))
            res = max(res, 1 + dfs(r - 1, c, matrix[r][c]))
            res = max(res, 1 + dfs(r, c + 1, matrix[r][c]))
            res = max(res, 1 + dfs(r, c - 1, matrix[r][c]))
            
            dp[(r, c)] = res
            return res
        
        for i in range(rows):
            for j in range(cols):
                dfs(i, j, -1)
        
        return max(dp.values())